dotnet/src/Extensions/PromptTemplate.Handlebars/Helpers/SystemHelpers/KernelSystemHelpers.cs

```diff
@@ -0,0 +1,177 @@
+    /// <exception cref="KernelException"></exception>
+    private static void RegisterSystemHelpers(
+        IHandlebars handlebarsInstance,
+        KernelArguments variables
+    )
+    {
+        // TODO [@teresaqhoang]: Issue 3947 Isolate Handlebars Kernel System helpers in their own class
+        handlebarsInstance.RegisterHelper("message", (writer, options, context, arguments) =>
+        {
+            var parameters = arguments[0] as IDictionary<string, object>;
+
+            // Verify that the message has a role
+            if (!parameters!.ContainsKey("role"))
+            {
+                throw new KernelException("Message must have a role.");
+            }
+
+            writer.Write($"<{parameters["role"]}~>", false);
+            options.Template(writer, context);
+            writer.Write($"</{parameters["role"]}~>", false);
+        });
+
+        handlebarsInstance.RegisterHelper("set", (writer, context, arguments) =>
+        {
+            var name = string.Empty;
+            object value = string.Empty;
+            if (arguments[0].GetType() == typeof(HashParameterDictionary))
+            {
+                // Get the parameters from the template arguments
+                var parameters = arguments[0] as IDictionary<string, object>;
+                name = (string)parameters!["name"];
+                value = parameters!["value"];
+            }
+            else
+            {
+                name = arguments[0].ToString();
+                value = arguments[1];
+            }
+
+            if (variables.TryGetValue(name, out var argVal))
+            {
+                variables[name] = value;
+            }
+            else
+            {
+                variables.Add(name, value);
+            }
+        });
+
+        handlebarsInstance.RegisterHelper("get", (in HelperOptions options, in Context context, in Arguments arguments) =>
+        {
+            if (arguments[0].GetType() == typeof(HashParameterDictionary))
+            {
+                var parameters = arguments[0] as IDictionary<string, object>;
+                return variables[(string)parameters!["name"]];
+            }
+
+            return variables[arguments[0].ToString()];
+        });
+
+        handlebarsInstance.RegisterHelper("json", (in HelperOptions options, in Context context, in Arguments arguments) =>
+        {
+            object objectToSerialize = arguments[0];
+            string json = objectToSerialize.GetType() == typeof(string) ? (string)objectToSerialize : JsonSerializer.Serialize(objectToSerialize);
+
+            return json;
+        });
+
+        handlebarsInstance.RegisterHelper("concat", (in HelperOptions options, in Context context, in Arguments arguments) =>
+        {
+            List<string?> strings = arguments.Select((var) =>
+            {
+                if (var == null)
+                {
+                    return null;
+                }
+                return var!.ToString();
```

Comment:
```
return var == null ? null : var.ToString();
```

Reply:
This should work as well: `return var?.ToString();`
All block in `RegisterHelper` could look like `return string.Concat(arguments.Select(var => var?.ToString()));`

Reply:
Love the optimizations suggestions, have been looking at this code for too long. Thank you!